File: appkillermanager/src/main/java/com/thelittlefireman/appkillermanager/devices/Asus.java
package com.thelittlefireman.appkillermanager.devices;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;

import com.thelittlefireman.appkillermanager.utils.ActionsUtils;
import com.thelittlefireman.appkillermanager.utils.Manufacturer;

// TODO TESTS
public class Asus extends DeviceAbstract {

    private static final String ASUS_PACAKGE_MOBILEMANAGER = "com.asus.mobilemanager";
    private static final String ASUS_ACTIVITY_MOBILEMANAGER_FUNCTION_ACTIVITY = "com.asus.mobilemanager.entry.FunctionActivity";
    private static final String ASUS_ACTIVITY_MOBILEMANAGER_FUNCTION_AUTOSTART_ACTIVITY = "com.asus.mobilemanager.autostart.AutoStartActivity";

    @Override
    public boolean isThatRom() {
        return  Build.BRAND.equalsIgnoreCase(getDeviceManufacturer().toString()) ||
                Build.MANUFACTURER.equalsIgnoreCase(getDeviceManufacturer().toString()) ||
                Build.FINGERPRINT.toLowerCase().contains(getDeviceManufacturer().toString());
    }

    @Override
    public Manufacturer getDeviceManufacturer() {
        return Manufacturer.ASUS;
    }

    @Override
    public boolean isActionPowerSavingAvailable(Context context) {
        return false;
    }

    @Override
    public boolean isActionAutoStartAvailable(Context context) {
        return true;
    }

    @Override
    public boolean isActionNotificationAvailable(Context context) {
        return false;
    }

    @Override
    public Intent getActionPowerSaving(Context context) {
        // TODO JUST NEED TO USE THE REGULAR BATTERY NON OPTIMISATION PERMISSION =)
        return null;
    }

    @Override
    public Intent getActionAutoStart(Context context) {
        Intent intent = ActionsUtils.createIntent();
        intent.putExtra("showNotice",true);
        intent.setComponent(new ComponentName(ASUS_PACAKGE_MOBILEMANAGER, ASUS_ACTIVITY_MOBILEMANAGER_FUNCTION_AUTOSTART_ACTIVITY));
        return intent;
    }

    @Override
    public Intent getActionNotification(Context context) {
        return null;
    }

    @Override
    public String getExtraDebugInformations(Context context) {
        return null;
    }

    @Override
    public int getHelpImage() {
        return 0;
    }
}

File: appkillermanager/src/main/java/com/thelittlefireman/appkillermanager/devices/Asus.kt
package com.thelittlefireman.appkillermanager.devices

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.PowerManager
import android.provider.Settings
import com.thelittlefireman.appkillermanager.R
import com.thelittlefireman.appkillermanager.utils.ActionsUtils
import com.thelittlefireman.appkillermanager.utils.Manufacturer
import timber.log.Timber

// TODO TESTS
class Asus : Device {

    companion object {
        private const val ASUS_PACKAGE_MOBILE_MANAGER = "com.asus.mobilemanager"
        private const val ASUS_ACTIVITY_MOBILE_MANAGER_FUNCTION_ACTIVITY = "com.asus.mobilemanager.entry.FunctionActivity"
        private const val ASUS_ACTIVITY_MOBILE_MANAGER_FUNCTION_AUTO_START_ACTIVITY = "com.asus.mobilemanager.autostart.AutoStartActivity"
    }

    override val isThatRom: Boolean
        get() = Build.BRAND.equals(manufacturer.toString(), ignoreCase = true) ||
                Build.MANUFACTURER.equals(manufacturer.toString(), ignoreCase = true) ||
                Build.FINGERPRINT.contains(manufacturer.toString(), ignoreCase = true)

    override val manufacturer: Manufacturer
        get() = Manufacturer.ASUS

    override fun isActionPowerSavingAvailable(context: Context): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
            return pm.isIgnoringBatteryOptimizations(context.packageName)
        }
        return false
    }

    override fun isActionAutoStartAvailable(context: Context) = true

    override fun isActionNotificationAvailable(context: Context) = true

    override fun getActionPowerSaving(context: Context): Intent? {
        // Just need to use the regular battery non optimization permission =)
        //Android 7.0+ Doze
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
            if (!pm.isIgnoringBatteryOptimizations(context.packageName)) {
                // Cannot fire Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS
                // due to Google play device policy restriction !
                return ActionsUtils.createIntent().apply {
                    action = Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS
                    data = Uri.parse("package:" + context.packageName)
                }
            } else {
                Timber.tag(this.javaClass.name).i("getActionDozeMode: App is already enable to ignore doze battery optimization")
            }
        }
        return null
    }

    override fun getActionAutoStart(context: Context): Intent = ActionsUtils.createIntent().apply {
        putExtra("showNotice", true)
        component = ComponentName(ASUS_PACKAGE_MOBILE_MANAGER, ASUS_ACTIVITY_MOBILE_MANAGER_FUNCTION_AUTO_START_ACTIVITY)
    }

    // Need to click on notifications items
    override fun getActionNotification(context: Context): Intent = ActionsUtils.createIntent().apply {
        putExtra("showNotice", true)
        component = ComponentName(ASUS_PACKAGE_MOBILE_MANAGER, ASUS_ACTIVITY_MOBILE_MANAGER_FUNCTION_ACTIVITY)
    }
}
